converter.go

```diff
@@ -354,6 +354,74 @@ func decimalToBigFloat(num decimal128.Num, scale int64) *big.Float {
 	return new(big.Float).Quo(f, s)
 }
 
+// ArrowSnowflakeTimestampToTime converts original timestamp returned by Snowflake to time.TIme
+func ArrowSnowflakeTimestampToTime(rec arrow.Record, rb *ArrowBatch, colIdx int, recIdx int) *time.Time {
+	scale := int(rb.scd.RowSet.RowType[colIdx].Scale)
+	dbType := strings.ToUpper(rb.scd.RowSet.RowType[colIdx].Type)
+	return arrowSnowflakeTimestampToTime(rec.Column(colIdx), getSnowflakeType(dbType), scale, recIdx, rb.loc)
+}
+
+func arrowSnowflakeTimestampToTime(
+	column arrow.Array,
+	sfType snowflakeType,
+	scale int,
+	recIdx int,
+	loc *time.Location) *time.Time {
+
+	if column.IsNull(recIdx) {
+		return nil
+	}
+	var ret time.Time
+	switch sfType {
+	case timestampNtzType:
+		if column.DataType().ID() == arrow.STRUCT {
+			structData := column.(*array.Struct)
+			epoch := structData.Field(0).(*array.Int64).Int64Values()
+			fraction := structData.Field(1).(*array.Int32).Int32Values()
+			ret = time.Unix(epoch[recIdx], int64(fraction[recIdx])).UTC()
+		} else {
+			intData := column.(*array.Int64)
+			value := intData.Value(recIdx)
+			epoch := value / int64(math.Pow10(scale))
+			fraction := (value % int64(math.Pow10(scale))) * int64(math.Pow10(9-scale))
+			ret = time.Unix(epoch, fraction).UTC()
+		}
+		return &ret
+	case timestampLtzType:
+		if column.DataType().ID() == arrow.STRUCT {
+			structData := column.(*array.Struct)
+			epoch := structData.Field(0).(*array.Int64).Int64Values()
+			fraction := structData.Field(1).(*array.Int32).Int32Values()
+			ret = time.Unix(epoch[recIdx], int64(fraction[recIdx])).In(loc)
+		} else {
+			intData := column.(*array.Int64)
+			value := intData.Value(recIdx)
+			epoch := value / int64(math.Pow10(scale))
+			fraction := (value % int64(math.Pow10(scale))) * int64(math.Pow10(9-scale))
+			ret = time.Unix(epoch, fraction).In(loc)
+		}
+		return &ret
+	case timestampTzType:
+		structData := column.(*array.Struct)
+		if structData.NumField() == 2 {
+			value := structData.Field(0).(*array.Int64).Int64Values()
+			timezone := structData.Field(1).(*array.Int32).Int32Values()
+			loc := Location(int(timezone[recIdx]) - 1440)
+			epoch := value[recIdx] / int64(math.Pow10(scale))
+			fraction := value[recIdx] % int64(math.Pow10(scale)) * int64(math.Pow10(9-scale))
```

Comment:
perhaps we can extract this to a method `extractFraction(value, scale)` because there are some repetitions of `int64(math.Pow10(scale)) * int64(math.Pow10(9-scale))`. Same with epoch